File: Dopamine.SettingsModule/ViewModels/SettingsViewModel.vb
Imports Microsoft.Practices.Prism.Mvvm
Imports Microsoft.Practices.Prism.Regions
Imports Microsoft.Practices.Prism.Commands
Imports Dopamine.Core.Prism
Imports Microsoft.Practices.Prism
Imports Dopamine.Common.Services.Indexing
Imports Dopamine.Core.Settings
Imports Dopamine.Common.Services.Collection

Namespace ViewModels
    Public Class SettingsViewModel
        Inherits BindableBase
        Implements IActiveAware
        Implements INavigationAware


#Region "Variables"
        Private ReadOnly mRegionManager As IRegionManager
        Private mIndexingService As IIndexingService
        Private mcollectionservice As ICollectionService
        Private mPreviousIndex As Integer = 0
        Private mSlideInFrom As Integer
#End Region

#Region "Properties"
        Public NavigateBetweenSettingsCommand As DelegateCommand(Of String)

        Public Property SlideInFrom() As Integer
            Get
                Return mSlideInFrom
            End Get
            Set(ByVal value As Integer)
                SetProperty(Of Integer)(mSlideInFrom, value)
            End Set
        End Property
#End Region

#Region "Construction"
        Public Sub New(iRegionManager As IRegionManager, iIndexingService As IIndexingService, iCollectionService As ICollectionService)
            mRegionManager = iRegionManager
            mIndexingService = iIndexingService
            mcollectionservice = iCollectionService
            Me.NavigateBetweenSettingsCommand = New DelegateCommand(Of String)(AddressOf NavigateBetweenSettings)
            ApplicationCommands.NavigateBetweenSettingsCommand.RegisterCommand(Me.NavigateBetweenSettingsCommand)
            Me.SlideInFrom = 30
        End Sub
#End Region

#Region "Events"
        Public Event IsActiveChanged(sender As Object, e As EventArgs) Implements IActiveAware.IsActiveChanged
#End Region

#Region "Private"
        Private Sub NavigateBetweenSettings(iIndex As String)

            If String.IsNullOrWhiteSpace(iIndex) Then Return

            Dim index As Integer = 0

            Integer.TryParse(iIndex, index)

            If index = 0 Then Return

            Me.SlideInFrom = If(index <= mPreviousIndex, -30, 30)

            mPreviousIndex = index

            mRegionManager.RequestNavigate(RegionNames.SettingsRegion, Me.GetPageForIndex(index))
        End Sub

        Private Function GetPageForIndex(iIndex As Integer) As String

            Dim page As String = String.Empty

            Select Case iIndex
                Case 1
                    page = GetType(Views.SettingsCollection).FullName
                Case 2
                    page = GetType(Views.SettingsAppearance).FullName
                Case 3
                    page = GetType(Views.SettingsBehaviour).FullName
                Case 4
                    page = GetType(Views.SettingsPlayback).FullName
                Case 5
                    page = GetType(Views.SettingsStartup).FullName
                Case Else
                    page = GetType(Views.SettingsCollection).FullName
            End Select

            Return page
        End Function
#End Region

#Region "Public"
        Public Function IsNavigationTarget(navigationContext As NavigationContext) As Boolean Implements INavigationAware.IsNavigationTarget
            Return True
        End Function

        Public Sub OnNavigatedFrom(navigationContext As NavigationContext) Implements INavigationAware.OnNavigatedFrom
            Me.mIndexingService.IndexCollectionAsync(XmlSettingsClient.Instance.Get(Of Boolean)("Indexing", "IgnoreRemovedFiles"), False)
            mcollectionservice.SaveMarkedFoldersAsync()
        End Sub

        Public Sub OnNavigatedTo(navigationContext As NavigationContext) Implements INavigationAware.OnNavigatedTo

        End Sub

        Public Property IsActive As Boolean Implements IActiveAware.IsActive
#End Region

    End Class
End Namespace

File: Dopamine.SettingsModule/ViewModels/SettingsViewModel.cs
using Digimezzo.Utilities.Settings;
using Dopamine.Common.Services.Collection;
using Dopamine.Common.Services.Indexing;
using Dopamine.Common.Prism;
using Prism;
using Prism.Commands;
using Prism.Mvvm;
using Prism.Regions;
using System;

namespace Dopamine.SettingsModule.ViewModels
{
    public class SettingsViewModel : BindableBase, IActiveAware, INavigationAware
    {
        #region Variables
        private readonly IRegionManager regionManager;
        private IIndexingService indexingService;
        private ICollectionService collectionService;
        private int previousIndex = 0;
        private int slideInFrom;
        #endregion

        #region Commands
        public DelegateCommand<string> NavigateBetweenSettingsCommand;
        #endregion

        #region Properties
        public bool IsActive { get; set; }

        public int SlideInFrom
        {
            get { return this.slideInFrom; }
            set { SetProperty<int>(ref this.slideInFrom, value); }
        }
        #endregion

        #region Construction
        public SettingsViewModel(IRegionManager regionManager, IIndexingService indexingService, ICollectionService collectionService)
        {
            this.regionManager = regionManager;
            this.indexingService = indexingService;
            this.collectionService = collectionService;
            this.NavigateBetweenSettingsCommand = new DelegateCommand<string>(NavigateBetweenSettings);
            ApplicationCommands.NavigateBetweenSettingsCommand.RegisterCommand(this.NavigateBetweenSettingsCommand);
            this.SlideInFrom = 30;
        }
        #endregion

        #region IActiveAware
        public event EventHandler IsActiveChanged;
        #endregion

        #region Private
        private void NavigateBetweenSettings(string iIndex)
        {
            if (string.IsNullOrWhiteSpace(iIndex))
                return;

            int index = 0;

            int.TryParse(iIndex, out index);

            if (index == 0)
                return;

            this.SlideInFrom = index <= this.previousIndex ? -30 : 30;

            this.previousIndex = index;

            this.regionManager.RequestNavigate(RegionNames.SettingsRegion, this.GetPageForIndex(index));
        }

        private string GetPageForIndex(int iIndex)
        {

            string page = string.Empty;

            switch (iIndex)
            {
                case 1:
                    page = typeof(Views.SettingsCollection).FullName;
                    break;
                case 2:
                    page = typeof(Views.SettingsAppearance).FullName;
                    break;
                case 3:
                    page = typeof(Views.SettingsBehaviour).FullName;
                    break;
                case 4:
                    page = typeof(Views.SettingsPlayback).FullName;
                    break;
                case 5:
                    page = typeof(Views.SettingsStartup).FullName;
                    break;
                case 6:
                    page = typeof(Views.SettingsOnline).FullName;
                    break;
                default:
                    page = typeof(Views.SettingsCollection).FullName;
                    break;
            }

            return page;
        }
        #endregion

        #region INavigationAware
        public bool IsNavigationTarget(NavigationContext navigationContext)
        {
            return true;
        }

        public void OnNavigatedFrom(NavigationContext navigationContext)
        {
            this.indexingService.CheckCollectionAsync();
            this.collectionService.SaveMarkedFoldersAsync();
        }

        public void OnNavigatedTo(NavigationContext navigationContext)
        {
        }
        #endregion
    }
}
